src/uct/ib/ud/base/ud_ep.c

```diff
@@ -1052,33 +1058,24 @@ ucs_status_t uct_ud_ep_flush_nolock(uct_ud_iface_t *iface, uct_ud_ep_t *ep,
     }
 
     /* Expedite acknowledgment on the last skb in the window */
-    if (uct_ud_ep_is_last_ack_received(ep)) {
+    if (uct_ud_ep_is_last_ack_received(ep) &&
+        !(ep->flags & UCT_UD_EP_FLAG_FLUSH_CANCEL)) {
         uct_ud_ep_ctl_op_del(ep, UCT_UD_EP_OP_ACK_REQ);
     } else {
-        uct_ud_ep_assert_tx_window_nonempty(ep);
-        skb = ucs_queue_tail_elem_non_empty(&ep->tx.window, uct_ud_send_skb_t,
-                                            queue);
-        if (!(skb->flags & UCT_UD_SEND_SKB_FLAG_ACK_REQ)) {
-            /* If we didn't ask for ACK on last skb, send an ACK_REQ message.
-             * It will speed up the flush because we will not have to wait until
-             * retransmit is triggered.
-             * Also, prevent from sending more control messages like this after
-             * first time by turning on the flag on the last skb.
-             */
-
-            /* Since the function can be called from the arbiter context it is
-             * impossible to schedule a control operation. So just raise a
-             * flag and if there is no other control send ACK_REQ directly.
-             *
-             * If there is other control arbiter will take care of it.
-             */
-            ep->tx.pending.ops |= UCT_UD_EP_OP_ACK_REQ;
-            if (uct_ud_ep_ctl_op_check_ex(ep, UCT_UD_EP_OP_ACK_REQ)) {
-                uct_ud_ep_do_pending_ctl(ep, iface);
-            }
+        /* Send an ACK_REQ message. It will speed up the flush because we will
+         * not have to wait until retransmit is triggered.
+         * Also, prevent from sending more control messages like this after
+         * first time by turning on the flag on the last skb.
```

Comment:
why removed the logic of UCT_UD_SEND_SKB_FLAG_ACK_REQ, that prevented sending multiple consecutive ACK packets?

Reply:
returned back for flush(LOCAL) case.
we shouldn't check the flag for flush(CANCEL), because we should ensure that we send ACK_REQ with `SIGNALED` flag set.

Reply:
maybe flush(LOCAL) should also send with signaled?

Reply:
done